src/core/internal/backtrace/dwarf.d

```diff
@@ -169,65 +169,90 @@ int traceHandlerOpApplyImpl(const(void*)[] callstack, scope int delegate(ref siz
 
     auto image = Image.openSelf();
 
-    int processCallstack(const(ubyte)[] debugLineSectionData)
+    // find address -> file, line mapping using dwarf debug_line
+    Array!Location locations;
+    locations.length = callstack.length;
+    foreach (size_t i; 0 .. callstack.length)
     {
-        // find address -> file, line mapping using dwarf debug_line
-        Array!Location locations;
-        locations.length = callstack.length;
-        foreach (size_t i; 0 .. callstack.length)
-        {
-            locations[i].address = callstack[i];
-            locations[i].procedure = getMangledSymbolName(frameList[i][0 .. strlen(frameList[i])]);
-        }
-
-        if (debugLineSectionData)
-            resolveAddresses(debugLineSectionData, locations[], image.baseAddress);
+        locations[i].address = callstack[i];
+        locations[i].procedure = getMangledSymbolName(frameList[i][0 .. strlen(frameList[i])]);
+    }
 
-        char[1536] buffer = void;
-        size_t bufferLength;
-        scope sink = (scope const char[] data) {
-                // We cannot write anymore
-                if (bufferLength > buffer.length)
-                    return;
+    return image.isValid
+        ? image.processDebugLineSectionData(
+            (line) => locations[].processCallstack(line, image.baseAddress, dg))
+        : locations[].processCallstack(null, image.baseAddress, dg);
+}
 
-                if (bufferLength + data.length > buffer.length)
-                {
-                    buffer[bufferLength .. $] = data[0 .. buffer.length - bufferLength];
-                    buffer[$ - 3 .. $] = "...";
-                    // +1 is a marker for the '...', otherwise if the symbol
-                    // name was to exactly fill the buffer,
-                    // we'd discard anything else without printing the '...'.
-                    bufferLength = buffer.length + 1;
-                    return;
-                }
+struct TraceInfoBuffer
+{
+    private char[1536] buf = void;
+    private size_t position;
 
-                buffer[bufferLength .. bufferLength + data.length] = data;
-                bufferLength += data.length;
-        };
+    // BUG: https://issues.dlang.org/show_bug.cgi?id=21285
+    @safe pure nothrow @nogc
+    {
+        ///
+        public inout(char)[] opSlice () inout return
+        {
+            return this.buf[0 .. this.position > $ ? $ : this.position];
+        }
 
-        foreach (idx, const ref loc; locations)
+        ///
+        public void reset ()
         {
-            bufferLength = 0;
-            loc.toString(sink);
+            this.position = 0;
+        }
+    }
 
-            auto lvalue = buffer[0 .. bufferLength > $ ? $ : bufferLength];
-            if (auto ret = dg(idx, lvalue))
-                return ret;
+    /// Used as `sink` argument to `Location.toString`
+    public void write(scope const char[] data)
```

Comment:
```suggestion
    void put(scope const char[] data)
```

To comply with the `OutputRange` interface: https://dlang.org/phobos/std_range_primitives.html#isOutputRange